utils/awsutils/session.go

```diff
@@ -0,0 +1,81 @@
+package awsutils
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+type SessionConfig struct {
+	Region      string
+	AccessKeyID string
+	AccessKey   string
+	IAMRoleARN  string
+	ExternalID  string
+	Service     string
+	Timeout     time.Duration
+}
+
+func createRoleSessionName(serviceName string) string {
+	return fmt.Sprintf("rudderstack-aws-%s-access", strings.ToLower(serviceName))
+}
+
+func createDefaultSession(config *SessionConfig) *session.Session {
+	return session.Must(session.NewSession(&aws.Config{
+		HTTPClient: &http.Client{
+			Timeout: config.Timeout,
+		},
+		Region: aws.String(config.Region),
+	}))
+}
+
+func createCredentailsForRole(config *SessionConfig) *credentials.Credentials {
+	hostSession := createDefaultSession(config)
+	return stscreds.NewCredentials(hostSession, config.IAMRoleARN,
+		func(p *stscreds.AssumeRoleProvider) {
+			p.ExternalID = aws.String(config.ExternalID)
+			p.RoleSessionName = createRoleSessionName(config.Service)
+		})
+}
+
+func createCredentails(config *SessionConfig) *credentials.Credentials {
+	if config.IAMRoleARN != "" {
+		return createCredentailsForRole(config)
+	} else if config.AccessKey != "" && config.AccessKeyID != "" {
+		return credentials.NewStaticCredentials(config.AccessKeyID, config.AccessKey, "")
+	}
+	return nil
+}
+
+func CreateSession(config *SessionConfig) *session.Session {
+	return session.Must(session.NewSession(&aws.Config{
+		HTTPClient: &http.Client{
+			Timeout: config.Timeout,
+		},
+		Region:      aws.String(config.Region),
+		Credentials: createCredentails(config),
+	}))
+}
+
+func NewSessionConfig(destinationConfig interface{}, timeout time.Duration, serviceName string) (*SessionConfig, error) {
+	sessionConfig := SessionConfig{}
+
+	jsonConfig, err := json.Marshal(destinationConfig)
```

Comment:
Any reason why we would need to do both json marshal and unmarshal instead of just type checking the input interface?

Reply:
Downstream components is sending destinationConfig as interface{} and it cannot be type casted to SessionConfig but it has same fields as session config so marshalling to json and then unmarshalling to SessionConfig, actually this is an existing code just moved here. Ideal would be taking destination config as map[string]interface{} and manually extract required fields but this itself is bit long change as we need to write custom builders for each time from map type so currently not planning to address as part of this PR.